packages/main/src/Calendar.ts

```diff
@@ -285,41 +294,104 @@ class Calendar extends CalendarPart {
-	}
+		let selectedDates: Array<number> = [];
 
-	constructor() {
-		super();
+		switch (this.selectionMode) {
+		case CalendarSelectionMode.Range: {
+			const range = this.dates.filter(date => date.hasAttribute("ui5-date-range"))[0];
+			const startDate = range && range.startValue && this.getFormat().parse(range.startValue, true) as Date;
+			const endDate = range && range.endValue && this.getFormat().parse(range.endValue, true) as Date;
 
-		this._valueIsProcessed = false;
+			if (startDate) {
+				selectedDates.push(startDate.getTime() / 1000);
+			}
+
+			if (endDate) {
+				selectedDates.push(endDate.getTime() / 1000);
+			}
+			break;
+		}
+		case CalendarSelectionMode.Multiple:
+		default: {
+			selectedDates = this.dates
+				.filter(dateElement => {
+					return dateElement.hasAttribute("ui5-date")
+						&& dateElement.value
+						&& this._isValidCalendarDate(dateElement.value)
+						&& this._getTimeStampFromString(dateElement.value);
+				})
+				.map(dateElement => Number(this._getTimeStampFromString(dateElement.value!)) / 1000);
+			break;
+		}
+		}
+
+		return selectedDates;
 	}
 
 	/**
 	 * @private
 	 */
 	_setSelectedDates(selectedDates: Array<number>) {
-		const selectedValues = selectedDates.map(timestamp => this.getFormat().format(UI5Date.getInstance(timestamp * 1000), true)); // Format as UTC
-		const valuesInDOM = [...this.dates].map(dateElement => dateElement.value);
-
-		// Remove all elements for dates that are no longer selected
-		this.dates.filter(dateElement => !selectedValues.includes(dateElement.value)).forEach(dateElement => {
-			this.removeChild(dateElement);
-		});
-
-		// Create tags for the selected dates that don't already exist in DOM
-		selectedValues.filter(value => !valuesInDOM.includes(value)).forEach(value => {
-			const dateElement = document.createElement(CalendarDate.getMetadata().getTag()) as CalendarDate;
-			dateElement.value = value;
-			this.appendChild(dateElement);
-		});
+		const selectedUTCDates = selectedDates.map(timestamp => this.getFormat().format(UI5Date.getInstance(timestamp * 1000), true));
+
+		switch (this.selectionMode) {
+		case CalendarSelectionMode.Range: {
+			// Create tags for the selected dates that don't already exist in DOM
+			if (selectedUTCDates.length) {
+				let dateRange = this.dates.find(dateElement => dateElement.startValue === selectedUTCDates[0]) as CalendarDateRange;
+				if (!dateRange) {
+					dateRange = document.createElement(CalendarDateRange.getMetadata().getTag()) as CalendarDateRange;
+					dateRange.startValue = selectedUTCDates[0];
+					this.appendChild(dateRange);
+				} else {
+					dateRange.endValue = selectedUTCDates[1];
+				}
+				// Remove all elements for dates that are no longer selected
+				this.dates
+					.filter(dateElement => {
+						return dateElement.hasAttribute("ui5-date") || dateElement.startValue !== dateRange.startValue;
+					})
+					.forEach(dateElement => {
+						this.removeChild(dateElement);
```

Comment:
Do we need to remove these childs?

Reply:
We want to keep the slot updated with the currently selected dates so we remove any dates no longer selected. Otherwise we can't really make new selections.